### ISSUE TITLE
[0.17] dependency resolution is incompatible with 0.16.1

### ISSUE THREAD

**Issue by bestander** [yarnpkg/yarn#1933]
**Do you want to request a *feature* or report a *bug*?**
Bug

**What is the current behavior?**

Installation done by yarn 0.16.1 is not compatible with 0.17.4.

```
bestander-mbp:react-native bestander$ yarn install
yarn install v0.16.1
[1/4] 🔍  Resolving packages...
[2/4] 🚚  Fetching packages...
[3/4] 🔗  Linking dependencies...
[4/4] 📃  Building fresh packages...
success Saved lockfile.
✨  Done in 30.91s.
bestander-mbp:react-native bestander$ yarn check
yarn check v0.16.1
warning babel-traverse#invariant@^2.2.0 could be deduped from 2.2.2 to invariant@2.2.2
success Folder in sync.
✨  Done in 1.93s.
bestander-mbp:react-native bestander$ node ~/Downloads/yarn-0.17.4.js check
yarn check v0.17.4
warning babel-traverse#invariant@^2.2.0 could be deduped from 2.2.2 to invariant@2.2.2
error react-redux#invariant is wrong version: expected 2.2.1, got 2.2.2
error optimist#minimist is wrong version: expected 0.0.8, got 0.0.10
error finalhandler#escape-html is wrong version: expected 1.0.2, got 1.0.3
error errorhandler#escape-html not installed
error serve-static#escape-html not installed
error serve-index#escape-html not installed
error body-parser#iconv-lite is wrong version: expected 0.4.11, got 0.4.13
error encoding#iconv-lite not installed
error are-we-there-yet#readable-stream is wrong version: expected 2.0.6, got 2.2.2
error formatio#samsam is wrong version: expected 1.1.2, got 1.1.3
error multipipe#duplexer2 is wrong version: expected 0.0.2, got 0.1.4
error got#duplexer2 not installed
error stream-combiner2#duplexer2 not installed
error send#escape-html not installed
error raw-body#iconv-lite not installed
error jsdom#iconv-lite not installed
error htmlparser2#readable-stream is wrong version: expected 1.1.14, got 1.1.13
error duplexer2#readable-stream not installed
error whatwg-encoding#iconv-lite not installed
error duplexer2#readable-stream#isarray not installed
error Found 20 errors.
info Visit https://yarnpkg.com/en/docs/cli/check for documentation about this command.
```

And vice versa

```
bestander-mbp:react-native bestander$ rm -rf node_modules/
bestander-mbp:react-native bestander$ node ~/Downloads/yarn-0.17.4.js install
yarn install v0.17.4
[1/4] 🔍  Resolving packages...
[2/4] 🚚  Fetching packages...
[3/4] 🔗  Linking dependencies...
[4/4] 📃  Building fresh packages...
✨  Done in 25.26s.
bestander-mbp:react-native bestander$ node ~/Downloads/yarn-0.17.4.js check
yarn check v0.17.4
success Folder in sync.
✨  Done in 1.48s.
bestander-mbp:react-native bestander$ yarn check
yarn check v0.16.1
error react-redux#invariant is wrong version: expected 2.2.2, got 2.2.1
error babel-traverse#invariant not installed
error redux#lodash not installed
error optimist#minimist is wrong version: expected 0.0.10, got 0.0.8
error serve-static#escape-html is wrong version: expected 1.0.3, got 1.0.2
error finalhandler#escape-html not installed
error encoding#iconv-lite is wrong version: expected 0.4.13, got 0.4.11
error body-parser#iconv-lite not installed
error type-is#mime-types not installed
error serve-index#mime-types not installed
error are-we-there-yet#readable-stream is wrong version: expected 2.2.2, got 2.0.6
error through2#readable-stream not installed
error concat-stream#readable-stream not installed
error formatio#samsam is wrong version: expected 1.1.3, got 1.1.2
error isomorphic-fetch#whatwg-fetch not installed
error readable-stream#buffer-shims not installed
error got#duplexer2 is wrong version: expected 0.1.4, got 0.0.2
error multipipe#duplexer2 not installed
error serve-index#mime-types#mime-db not installed
error type-is#mime-types#mime-db not installed
error errorhandler#accepts#mime-types not installed
error accepts#mime-types not installed
error htmlparser2#domelementtype not installed
error accepts#mime-types#mime-db not installed
error errorhandler#mime-types#mime-db not installed
error multipipe#duplexer2#readable-stream not installed
error htmlparser2#readable-stream is wrong version: expected 1.1.13, got 1.1.14
error jsdom#sax not installed
error domhandler#domelementtype not installed
error domutils#domelementtype not installed
error multipipe#readable-stream#isarray not installed
error request#mime-types not installed
error tar-stream#end-of-stream not installed
error request#mime-types#mime-db not installed
error tar-stream#end-of-stream#once not installed
error bl#readable-stream not installed
error Found 36 errors.
info Visit https://yarnpkg.com/en/docs/cli/check for documentation about this command.
```

(ignore that I use bundled build for 0.17.4, the behavior is the same when you check with installed package)

**Comment by moimael** [comment]
Yes after updating I have missing dependencies issue during my build and test phase.

**Comment by bestander** [comment]
Mael, can you provide an example how to repro?

**Comment by moimael** [comment]
https://gist.github.com/moimael/10ee8a4ff11c8e2a8ed31b87ada4b9b7

Try to use those dependencies in a package.json file with yarn.

Then try to run `NODE_ENV=production node --harmony ./node_modules/.bin/webpack -p `
You should get: `Error: Cannot find module 'optimist'`

Then try to run eslint and you'll get `Error: Cannot find module 'concat-stream'`

I'm using 0.17.4

**Comment by bestander** [comment]
@moimael, I can't reproduce the "Cannot find module" issues with your gist.
With yarn 0.17.4
```
$ yarn install
yarn install v0.17.4
info No lockfile found.
warning No license field
[1/4] 🔍  Resolving packages...
warning raven > node-uuid@1.4.7: use uuid module instead
warning jest-cli > istanbul-api > fileset > minimatch@2.0.10: Please update to minimatch 3.0.2 or higher to avoid a RegExp DoS issue
warning @kadira/storybook > request > node-uuid@1.4.7: use uuid module instead
[2/4] 🚚  Fetching packages...
[3/4] 🔗  Linking dependencies...
warning Unmet peer dependency "@kadira/storybook-addons@^v1.3.1".
warning Unmet peer dependency "@kadira/storybook-addons@^1.3.0".
[4/4] 📃  Building fresh packages...
success Saved lockfile.
✨  Done in 45.18s.

$ yarn check
yarn check v0.17.4
warning No license field
warning react-router#history@^4.3.0 could be deduped from 4.4.0 to history@4.4.0
error @kadira/storybook##postcss-loader@1.1.0 doesn't satisfy found match of postcss-loader@1.1.1
warning chokidar#fsevents#node-pre-gyp@^0.6.29 could be deduped from 0.6.31 to node-pre-gyp@0.6.31
error compressible#mime-db is wrong version: expected 1.24.0, got 1.25.0
warning webpack-config-options#npm#abbrev@~1.0.9 could be deduped from 1.0.9 to abbrev@1.0.9
warning webpack-config-options#npm#ansicolors@~0.3.2 could be deduped from 0.3.2 to ansicolors@0.3.2
warning webpack-config-options#npm#ansistyles@~0.1.3 could be deduped from 0.1.3 to ansistyles@0.1.3
warning webpack-config-options#npm#aproba@~1.0.4 could be deduped from 1.0.4 to aproba@1.0.4
warning webpack-config-options#npm#archy@~1.0.0 could be deduped from 1.0.0 to archy@1.0.0
warning webpack-config-options#npm#asap@~2.0.5 could be deduped from 2.0.5 to asap@2.0.5
warning webpack-config-options#npm#chownr@~1.0.1 could be deduped from 1.0.1 to chownr@1.0.1
warning webpack-config-options#npm#cmd-shim@~2.0.2 could be deduped from 2.0.2 to cmd-shim@2.0.2
warning webpack-config-options#npm#columnify@~1.5.4 could be deduped from 1.5.4 to columnify@1.5.4
warning webpack-config-options#npm#config-chain@~1.1.11 could be deduped from 1.1.11 to config-chain@1.1.11
warning webpack-config-options#npm#dezalgo@~1.0.3 could be deduped from 1.0.3 to dezalgo@1.0.3
warning webpack-config-options#npm#editor@~1.0.0 could be deduped from 1.0.0 to editor@1.0.0
warning webpack-config-options#npm#fs-vacuum@~1.2.9 could be deduped from 1.2.9 to fs-vacuum@1.2.9
warning webpack-config-options#npm#fs-write-stream-atomic@~1.0.8 could be deduped from 1.0.8 to fs-write-stream-atomic@1.0.8
warning webpack-config-options#npm#fstream@~1.0.10 could be deduped from 1.0.10 to fstream@1.0.10
warning webpack-config-options#npm#fstream-npm@~1.2.0 could be deduped from 1.2.0 to fstream-npm@1.2.0
warning webpack-config-options#npm#glob@~7.1.0 could be deduped from 7.1.1 to glob@7.1.1
warning webpack-config-options#npm#graceful-fs@~4.1.9 could be deduped from 4.1.10 to graceful-fs@4.1.10
warning webpack-config-options#npm#has-unicode@~2.0.1 could be deduped from 2.0.1 to has-unicode@2.0.1
warning webpack-config-options#npm#hosted-git-info@~2.1.5 could be deduped from 2.1.5 to hosted-git-info@2.1.5
warning webpack-config-options#npm#iferr@~0.1.5 could be deduped from 0.1.5 to iferr@0.1.5
warning webpack-config-options#npm#inflight@~1.0.5 could be deduped from 1.0.6 to inflight@1.0.6
warning webpack-config-options#npm#inherits@~2.0.3 could be deduped from 2.0.3 to inherits@2.0.3
warning webpack-config-options#npm#ini@~1.3.4 could be deduped from 1.3.4 to ini@1.3.4
warning webpack-config-options#npm#init-package-json@~1.9.4 could be deduped from 1.9.4 to init-package-json@1.9.4
warning webpack-config-options#npm#lockfile@~1.0.2 could be deduped from 1.0.2 to lockfile@1.0.2
warning webpack-config-options#npm#lodash._baseuniq@~4.6.0 could be deduped from 4.6.0 to lodash._baseuniq@4.6.0
warning webpack-config-options#npm#lodash.union@~4.6.0 could be deduped from 4.6.0 to lodash.union@4.6.0
warning webpack-config-options#npm#lodash.uniq@~4.5.0 could be deduped from 4.5.0 to lodash.uniq@4.5.0
warning webpack-config-options#npm#lodash.without@~4.4.0 could be deduped from 4.4.0 to lodash.without@4.4.0
warning webpack-config-options#npm#mkdirp@~0.5.1 could be deduped from 0.5.1 to mkdirp@0.5.1
warning webpack-config-options#npm#node-gyp@~3.4.0 could be deduped from 3.4.0 to node-gyp@3.4.0
warning webpack-config-options#npm#nopt@~3.0.6 could be deduped from 3.0.6 to nopt@3.0.6
warning webpack-config-options#npm#normalize-git-url@~3.0.2 could be deduped from 3.0.2 to normalize-git-url@3.0.2
warning webpack-config-options#npm#normalize-package-data@~2.3.5 could be deduped from 2.3.5 to normalize-package-data@2.3.5
warning webpack-config-options#npm#npm-cache-filename@~1.0.2 could be deduped from 1.0.2 to npm-cache-filename@1.0.2
warning webpack-config-options#npm#npm-install-checks@~3.0.0 could be deduped from 3.0.0 to npm-install-checks@3.0.0
warning webpack-config-options#npm#npm-package-arg@~4.2.0 could be deduped from 4.2.0 to npm-package-arg@4.2.0
warning webpack-config-options#npm#npm-registry-client@~7.2.1 could be deduped from 7.2.1 to npm-registry-client@7.2.1
warning webpack-config-options#npm#npm-user-validate@~0.1.5 could be deduped from 0.1.5 to npm-user-validate@0.1.5
warning webpack-config-options#npm#npmlog@~4.0.0 could be deduped from 4.0.1 to npmlog@4.0.1
warning webpack-config-options#npm#once@~1.4.0 could be deduped from 1.4.0 to once@1.4.0
warning webpack-config-options#npm#opener@~1.4.2 could be deduped from 1.4.2 to opener@1.4.2
warning webpack-config-options#npm#osenv@~0.1.3 could be deduped from 0.1.3 to osenv@0.1.3
warning webpack-config-options#npm#path-is-inside@~1.0.2 could be deduped from 1.0.2 to path-is-inside@1.0.2
warning webpack-config-options#npm#read@~1.0.7 could be deduped from 1.0.7 to read@1.0.7
warning webpack-config-options#npm#read-cmd-shim@~1.0.1 could be deduped from 1.0.1 to read-cmd-shim@1.0.1
warning webpack-config-options#npm#read-installed@~4.0.3 could be deduped from 4.0.3 to read-installed@4.0.3
warning webpack-config-options#npm#read-package-json@~2.0.4 could be deduped from 2.0.4 to read-package-json@2.0.4
warning webpack-config-options#npm#read-package-tree@~5.1.5 could be deduped from 5.1.5 to read-package-tree@5.1.5
warning webpack-config-options#npm#realize-package-specifier@~3.0.3 could be deduped from 3.0.3 to realize-package-specifier@3.0.3
warning webpack-config-options#npm#retry@~0.10.0 could be deduped from 0.10.0 to retry@0.10.0
warning webpack-config-options#npm#rimraf@~2.5.4 could be deduped from 2.5.4 to rimraf@2.5.4
warning webpack-config-options#npm#semver@~5.3.0 could be deduped from 5.3.0 to semver@5.3.0
warning webpack-config-options#npm#sha@~2.0.1 could be deduped from 2.0.1 to sha@2.0.1
warning webpack-config-options#npm#slide@~1.1.6 could be deduped from 1.1.6 to slide@1.1.6
warning webpack-config-options#npm#sorted-object@~2.0.1 could be deduped from 2.0.1 to sorted-object@2.0.1
warning webpack-config-options#npm#strip-ansi@~3.0.1 could be deduped from 3.0.1 to strip-ansi@3.0.1
warning webpack-config-options#npm#tar@~2.2.1 could be deduped from 2.2.1 to tar@2.2.1
warning webpack-config-options#npm#text-table@~0.2.0 could be deduped from 0.2.0 to text-table@0.2.0
warning webpack-config-options#npm#uid-number@0.0.6 could be deduped from 0.0.6 to uid-number@0.0.6
warning webpack-config-options#npm#umask@~1.1.0 could be deduped from 1.1.0 to umask@1.1.0
warning webpack-config-options#npm#unique-filename@~1.1.0 could be deduped from 1.1.0 to unique-filename@1.1.0
warning webpack-config-options#npm#unpipe@~1.0.0 could be deduped from 1.0.0 to unpipe@1.0.0
warning webpack-config-options#npm#validate-npm-package-name@~2.2.2 could be deduped from 2.2.2 to validate-npm-package-name@2.2.2
warning webpack-config-options#npm#which@~1.2.11 could be deduped from 1.2.12 to which@1.2.12
warning webpack-config-options#npm#wrappy@~1.0.2 could be deduped from 1.0.2 to wrappy@1.0.2
warning webpack-config-options#npm#write-file-atomic@~1.2.0 could be deduped from 1.2.0 to write-file-atomic@1.2.0
warning npm#readable-stream#buffer-shims@^1.0.0 could be deduped from 1.0.0 to buffer-shims@1.0.0
warning npm#readable-stream#core-util-is@~1.0.0 could be deduped from 1.0.2 to core-util-is@1.0.2
warning npm#readable-stream#inherits@~2.0.1 could be deduped from 2.0.3 to inherits@2.0.3
warning npm#readable-stream#isarray@~1.0.0 could be deduped from 1.0.0 to npm#isarray@1.0.0
warning npm#readable-stream#process-nextick-args@~1.0.6 could be deduped from 1.0.7 to process-nextick-args@1.0.7
warning npm#readable-stream#string_decoder@~0.10.x could be deduped from 0.10.31 to string_decoder@0.10.31
warning npm#readable-stream#util-deprecate@~1.0.1 could be deduped from 1.0.2 to util-deprecate@1.0.2
warning npm#request#aws-sign2@~0.6.0 could be deduped from 0.6.0 to aws-sign2@0.6.0
warning npm#request#aws4@^1.2.1 could be deduped from 1.5.0 to aws4@1.5.0
warning npm#request#bl@~1.1.2 could be deduped from 1.1.2 to bl@1.1.2
warning npm#request#caseless@~0.11.0 could be deduped from 0.11.0 to caseless@0.11.0
warning npm#request#combined-stream@~1.0.5 could be deduped from 1.0.5 to combined-stream@1.0.5
warning npm#request#extend@~3.0.0 could be deduped from 3.0.0 to npm#extend@3.0.0
warning npm#request#forever-agent@~0.6.1 could be deduped from 0.6.1 to forever-agent@0.6.1
warning npm#request#form-data@~2.0.0 could be deduped from 2.0.0 to npm#form-data@2.0.0
warning npm#request#har-validator@~2.0.6 could be deduped from 2.0.6 to har-validator@2.0.6
warning npm#request#hawk@~3.1.3 could be deduped from 3.1.3 to hawk@3.1.3
warning npm#request#http-signature@~1.1.0 could be deduped from 1.1.1 to http-signature@1.1.1
warning npm#request#is-typedarray@~1.0.0 could be deduped from 1.0.0 to is-typedarray@1.0.0
warning npm#request#isstream@~0.1.2 could be deduped from 0.1.2 to isstream@0.1.2
warning npm#request#json-stringify-safe@~5.0.1 could be deduped from 5.0.1 to json-stringify-safe@5.0.1
warning npm#request#mime-types@~2.1.7 could be deduped from 2.1.12 to mime-types@2.1.12
warning npm#request#node-uuid@~1.4.7 could be deduped from 1.4.7 to node-uuid@1.4.7
warning npm#request#oauth-sign@~0.8.1 could be deduped from 0.8.2 to oauth-sign@0.8.2
warning npm#request#stringstream@~0.0.4 could be deduped from 0.0.5 to stringstream@0.0.5
warning npm#request#tough-cookie@~2.3.0 could be deduped from 2.3.2 to tough-cookie@2.3.2
warning npm#request#tunnel-agent@~0.4.1 could be deduped from 0.4.3 to tunnel-agent@0.4.3
info Found 97 warnings.
error Found 2 errors.
info Visit https://yarnpkg.com/en/docs/cli/check for documentation about this command.

$ NODE_ENV=production node --harmony ./node_modules/.bin/webpack -p
webpack 1.13.3
Usage: https://webpack.github.io/docs/cli.html

Options:
  --help, -h, -?
  --config
  --context
  --entry
  --module-bind
  --module-bind-post
  --module-bind-pre
  --output-path
  --output-file
  --output-chunk-file
  --output-named-chunk-file
  --output-source-map-file
  --output-public-path
  --output-jsonp-function
  --output-pathinfo
  --output-library
  --output-library-target
  --records-input-path
  --records-output-path
  --records-path
  --define
  --target
  --cache                                                                                           [default: true]
  --watch, -w
  --watch which closes when stdin ends
  --watch-aggregate-timeout
  --watch-poll
  --hot
  --debug
  --devtool
  --progress
  --resolve-alias
  --resolve-loader-alias
  --optimize-max-chunks
  --optimize-min-chunk-size
  --optimize-minimize
  --optimize-occurence-order
  --optimize-dedupe
  --prefetch
  --provide
  --labeled-modules
  --plugin
  --bail
  --profile
  -d                                    shortcut for --debug --devtool sourcemap --output-pathinfo
  -p                                    shortcut for --optimize-minimize
  --json, -j
  --colors, -c
  --sort-modules-by
  --sort-chunks-by
  --sort-assets-by
  --hide-modules
  --display-exclude
  --display-modules
  --display-chunks
  --display-error-details
  --display-origins
  --display-cached
  --display-cached-assets
  --display-reasons, --verbose, -v

Output filename not configured.
```

Check shows some inconsistency but I can start webpack and eslint

**Comment by moimael** [comment]
That's weird I started it fresh (without lock file or node modules dir) several time and I can reproduce everytime. I am on Linux, not sure if that can be the cause

**Comment by bestander** [comment]
@moimael, this could be related to another bug we are crushing now related to caches.
Can you run `yarn cache clean` and reproduce?

**Comment by moimael** [comment]
I'll try this on Monday !

**Comment by bestander** [comment]
Interesting, some of the yarn check breakage is introduced by ec8ec8ac8b0eb459cf632cf26facb36615300124

**Comment by bestander** [comment]
The rest is fixed with reverting ec6a4508ad4aa748fc9ff9b2c665228bf5770696

**Comment by bestander** [comment]
I think the changes introduced in https://github.com/yarnpkg/yarn/commit/ec8ec8ac8b0eb459cf632cf26facb36615300124  https://github.com/yarnpkg/yarn/commit/ec6a4508ad4aa748fc9ff9b2c665228bf5770696 are legit and they can affect installation order.

In addition we need to introduce #1934 that should verify correctness of node_modules independent of sort or resolution implementation.

I'll close this task.
@moimael feel free to open a new issue regarding your case
